Add Docker

sashajaaa opened the pull request.

sashajaaa committed:
Commit №16.1:
feat:
-добавлен Docker.

sashajaaa committed:
Commit №16.1:
feat:
-добавлен Docker.

TyutterinYakov commented on gateway/pom.xml:
```diff
@@ -0,0 +1,70 @@
+        <dependency>
+            <groupId>org.springframework</groupId>
+            <artifactId>spring-webflux</artifactId>
+            <version>5.3.23</version>
```
Зависимости, для которых требуется указание версий следует располагать в блоке `<dependencyManagement>` в родительском POM, а в дочерних указывать их уже без ее указания
Но, в данном случае, более удобно было бы использовать `spring-boot-starter-webflux`, чтобы обеспечить полную совместимость

TyutterinYakov commented on gateway/src/main/java/ru/practicum/shareit/booking/BookingController.java:
```diff
@@ -0,0 +1,78 @@
+    @PostMapping
+    public ResponseEntity<Object> create(@RequestHeader(OWNER_ID_HEADER) long userId,
+                                         @RequestBody @Valid BookItemRequestDto requestDto) {
+        log.info("Creating booking {}, userId={}", requestDto, userId);
```
не забываем проверить, что время начала бронирования не равно времени окончания, а также то, что время начала не идет после времени окончания)
Если будет интересно, то эту валидацию можно было бы реализовать с помощью аннотации валидации над классом, чтобы поддерживать декларативный подход
https://devcolibri.com/spring-mvc-кастомная-аннотация-для-валидации/
Кстати, если будет интересно то, как реализовать кастомную аннотацию валидации на уровне класса `BookItemRequestDto` - можешь чекнуть тут) https://github.com/TyutterinYakov/CustomValidationClassLevel По примеру может быть легче разобраться с тем, что требуется)

TyutterinYakov commented on gateway/src/main/java/ru/practicum/shareit/exception/ErrorHandler.java:
```diff
@@ -0,0 +1,61 @@
+import javax.validation.ValidationException;
+@RestControllerAdvice
+@Slf4j
+public class ErrorHandler {
```
В каждый наш хендлер следует добавить логирование сообщения об ошибке)
Также хорошо было бы передать в лог сообщение об ошибке, а можно и весь трейс, чтобы в случае ошибки было проще разобраться, что именно произошло)
Кстати, при логгировании более эффективно использовать не конкатенацию, а заполнитель - placeholder `{}`
Например:
`log.debug("Получен статус 404 Not found {}", ex.getMessage(), ex);`
Аналогично в других местах)

TyutterinYakov commented on gateway/src/main/java/ru/practicum/shareit/item/ItemController.java:
```diff
@@ -0,0 +1,68 @@
+    @GetMapping("/search")
+    public ResponseEntity<Object> searchItems(@RequestHeader(OWNER_ID_HEADER) Long userId,
+                                              @RequestParam String text,
+                                              @RequestParam(value = "from", defaultValue = "0") Integer from,
```
Почему бы не использовать более декларативный подход и использовать аннотации ограничения) Например, для from можно было бы использовать `@PositiveOrZero`, так как значение может быть 0, либо больше) Аналогично и для size, но в этом случае следует использовать `@Positive`, так как значение не может быть равно 0
Кстати, для того, чтобы эти аннотации стали учитываться, необходимо над классом навесить аннотацию `@Validated`. Но выбрасываться будет не `MethodArgumentNotValidException`, к которому мы привыкли, а `ConstraintViolationException`

TyutterinYakov commented on gateway/src/main/java/ru/practicum/shareit/item/ItemController.java:
```diff
@@ -0,0 +1,68 @@
+
+    @GetMapping
+    public ResponseEntity<Object> getAllUsersItems(@RequestHeader(OWNER_ID_HEADER) Long userId,
+                                                   @RequestParam(value = "from", defaultValue = "0") Integer from,
```
Почему бы не использовать более декларативный подход и использовать аннотации ограничения) Например, для from можно было бы использовать `@PositiveOrZero`, так как значение может быть 0, либо больше) Аналогично и для size, но в этом случае следует использовать `@Positive`, так как значение не может быть равно 0
Кстати, для того, чтобы эти аннотации стали учитываться, необходимо над классом навесить аннотацию `@Validated`. Но выбрасываться будет не `MethodArgumentNotValidException`, к которому мы привыкли, а `ConstraintViolationException`

TyutterinYakov commented on gateway/src/main/java/ru/practicum/shareit/item/dto/CommentDto.java:
```diff
@@ -0,0 +1,23 @@
+@Data
+@AllArgsConstructor
+@NoArgsConstructor
+public class CommentDto {
```
При создании нам требуется передавать лишь текст комментария) Остальные же поля можно исключить)
Хорошо было бы ограничить размеры текстовых полей на основе тех ограничений, которые установлены на уровне таблицы в БД, чтобы заранее откинуть некорректные запросы) Для ограничения количества символов можно использовать `@Size`

TyutterinYakov commented on gateway/src/main/java/ru/practicum/shareit/item/dto/ItemDto.java:
```diff
@@ -0,0 +1,29 @@
+@Data
+@AllArgsConstructor
+@NoArgsConstructor
+public class ItemDto {
```
Хорошо было бы ограничить размеры текстовых полей на основе тех ограничений, которые установлены на уровне таблицы в БД, чтобы заранее откинуть некорректные запросы) Для ограничения количества символов можно использовать `@Size`

TyutterinYakov commented on gateway/src/main/java/ru/practicum/shareit/request/dto/ItemRequestDto.java:
```diff
@@ -0,0 +1,21 @@
+@Data
+@AllArgsConstructor
+@NoArgsConstructor
+public class ItemRequestDto {
```
Хорошо было бы ограничить размеры текстовых полей на основе тех ограничений, которые установлены на уровне таблицы в БД, чтобы заранее откинуть некорректные запросы) Для ограничения количества символов можно использовать `@Size`

TyutterinYakov commented on gateway/src/main/java/ru/practicum/shareit/user/UserController.java:
```diff
@@ -0,0 +1,50 @@
+    private final UserClient userClient;
+
+    @PostMapping
+    public ResponseEntity<Object> createUser(@Valid @RequestBody UserDto userDto) {
```
Кстати, вместо `@Valid` можно использовать такую аннотацию `@Validated(Create.class)`. Таким образом мы указываем, какую группу аннотаций следует учитывать. `Create` - это интерфейс, который ты создаешь самостоятельно, он будет пустым и служит лишь для указания группы) Аналогично следует поступить при обновлении, но там будет аннотация `@Validated(Update.class)`

Если не использовать валидацию по группам, тогда потребуется при обновлении дополнительно проверять формат переданной строки с почтой в коде, что не очень удобно. Либо же потребуется ввести отдельный класс DTO для обновления, чтобы над полем с почтой использовать `@Email`
https://struchkov.dev/blog/ru/spring-boot-validation/

TyutterinYakov commented on gateway/src/main/java/ru/practicum/shareit/user/dto/UserDto.java:
```diff
@@ -0,0 +1,26 @@
+@Data
+@AllArgsConstructor
+@NoArgsConstructor
+public class UserDto {
```
Так же и здесь, можно было бы использовать валидацию по группам)
Для имени:
`@NotBlank(groups = {Create.class})`
Для почты аннотацию `@Email` следует учитываться для обновления и создания, так как, если она не будет передана вообще, то все будет окей, а если будет, тогда она будет проверена на соответствие почте, а вот аннотацию
`@NotEmpty` только для создания) Так как при обновлении передавать ее не обязательно)

TyutterinYakov commented on server/pom.xml:
```diff
@@ -0,0 +1,90 @@
+			<scope>test</scope>
+		</dependency>
+        <dependency>
+            <groupId>jakarta.validation</groupId>
```
Зависимость с валидацей следует исключить, так как теперь вся валидация, которая не требует обращения к БД, должна находиться в модуле `gateway`

TyutterinYakov commented on gateway/src/main/java/ru/practicum/shareit/item/ItemController.java:
```diff
@@ -0,0 +1,68 @@
+                                              @RequestParam String text,
+                                              @RequestParam(value = "from", defaultValue = "0") Integer from,
+                                              @RequestParam(value = "size", defaultValue = "10") Integer size) {
+        return itemClient.searchItems(userId, text, from, size);
```
Почему бы сразу не проверить полученное значение `text` с помощью `isBlank`, и, если условие будет true, вернуть сразу пустую неизменяемую коллекцию, не обращаясь к серверу)

TyutterinYakov reviewed (commented):
Добрый вечер! В целом - хорошо) Отметил то, что следует исправить и то, что можно сделать лучше) Хорошего вечера! Если возникнут вопросы - пиши в пачку, буду рад помочь🌃

sashajaaa committed:
Commit №16.2:
refactor:
-переработан родительский файл pom.xml;
-переработан файл gateway/pom.xml;
-переработан класс gateway/BookingItemRequestDto;
-добавлен класс gateway/CheckDateValidator;
-добавлен интерфейс gateway/StartBeforeEndDateValid;
-переработан класс gateway/ErrorHandler;
-переработан класс gateway/CommentDto;
-переработан класс gateway/ItemDto;
-переработан класс gateway/ItemController;
-переработан класс gateway/ItemRequestDto;
-добавлен интерфейс gateway/Marker;
-переработан класс gateway/UserController;
-переработан файл server/pom.xml;
-переработан файл server/schema.sql;
-переработан класс server/BookingController;
-переработан класс server/InputBookingDto;
-переработан класс server/ErrorHandler;
-переработан класс server/CommentDto;
-переработан класс server/ItemController;
-переработан класс server/ItemDto;
-переработан класс server/ItemRequestController;
-переработан класс server/ItemRequestDto;
-переработан класс server/UserController;
-переработан класс server/UserDto.

sashajaaa committed:
Commit №16.2:
refactor:
-переработан родительский файл pom.xml;
-переработан файл gateway/pom.xml;
-переработан класс gateway/BookingItemRequestDto;
-добавлен класс gateway/CheckDateValidator;
-добавлен интерфейс gateway/StartBeforeEndDateValid;
-переработан класс gateway/ErrorHandler;
-переработан класс gateway/CommentDto;
-переработан класс gateway/ItemDto;
-переработан класс gateway/ItemController;
-переработан класс gateway/ItemRequestDto;
-добавлен интерфейс gateway/Marker;
-переработан класс gateway/UserController;
-переработан файл server/pom.xml;
-переработан файл server/schema.sql;
-переработан класс server/BookingController;
-переработан класс server/InputBookingDto;
-переработан класс server/ErrorHandler;
-переработан класс server/CommentDto;
-переработан класс server/ItemController;
-переработан класс server/ItemDto;
-переработан класс server/ItemRequestController;
-переработан класс server/ItemRequestDto;
-переработан класс server/UserController;
-переработан класс server/UserDto.

sashajaaa committed:
Commit №16.2:
refactor:
-переработан родительский файл pom.xml;
-переработан файл gateway/pom.xml;
-переработан класс gateway/BookingItemRequestDto;
-добавлен класс gateway/CheckDateValidator;
-добавлен интерфейс gateway/StartBeforeEndDateValid;
-переработан класс gateway/ErrorHandler;
-переработан класс gateway/CommentDto;
-переработан класс gateway/ItemDto;
-переработан класс gateway/ItemController;
-переработан класс gateway/ItemRequestDto;
-добавлен интерфейс gateway/Marker;
-переработан класс gateway/UserController;
-переработан файл server/pom.xml;
-переработан файл server/schema.sql;
-переработан класс server/BookingController;
-переработан класс server/InputBookingDto;
-переработан класс server/ErrorHandler;
-переработан класс server/CommentDto;
-переработан класс server/ItemController;
-переработан класс server/ItemDto;
-переработан класс server/ItemRequestController;
-переработан класс server/ItemRequestDto;
-переработан класс server/UserController;
-переработан класс server/UserDto.

TyutterinYakov commented on gateway/src/main/java/ru/practicum/shareit/booking/dto/BookItemRequestDto.java:
```diff
@@ -0,0 +1,30 @@
+    @NotNull
+    private long itemId;
+
+    @NotNull
```
Аннотация `@NotNull` не требуется, так как она уже реализована в кастомном валидаторе, который выполняется раньше, так как аннотация навешена над классом)

TyutterinYakov commented on gateway/src/main/java/ru/practicum/shareit/booking/dto/BookItemRequestDto.java:
```diff
@@ -0,0 +1,30 @@
+    @Future
+    private LocalDateTime start;
+
+    @NotNull
```
Аннотация `@NotNull` не требуется, так как она уже реализована в кастомном валидаторе, который выполняется раньше, так как аннотация навешена над классом)

TyutterinYakov commented on gateway/src/main/java/ru/practicum/shareit/booking/dto/BookItemRequestDto.java:
```diff
@@ -0,0 +1,30 @@
+    private LocalDateTime start;
+
+    @NotNull
+    @Future
```
Также аннотация `@Future` над `end` излишня, так как сначала будет проведена проверка в кастомной валидации, а после мы перейдем к проверке `start`, которая, если удовлетворяет `@FutureOrPresent`, тогда и `end` будет

TyutterinYakov commented on gateway/src/main/java/ru/practicum/shareit/booking/dto/BookItemRequestDto.java:
```diff
@@ -0,0 +1,30 @@
+    private long itemId;
+
+    @NotNull
+    @Future
```
Для даты старта хорошо было бы использовать не `@Future`, а `@FutureOrPresent`, так как дата старта может равняться текущему времени)

TyutterinYakov commented on gateway/src/main/java/ru/practicum/shareit/booking/dto/BookItemRequestDto.java:
```diff
@@ -0,0 +1,30 @@
+@StartBeforeEndDateValid
+public class BookItemRequestDto {
+
+    @NotNull
```
Аннотация `@NotNull` не используется над примитивами)

TyutterinYakov commented on gateway/src/main/java/ru/practicum/shareit/item/ItemController.java:
```diff
@@ -0,0 +1,75 @@
+                                              @PositiveOrZero @RequestParam(value = "from", defaultValue = "0") Integer from,
+                                              @Positive @RequestParam(value = "size", defaultValue = "10") Integer size) {
+        if (text.isBlank()) {
+            return new ResponseEntity<>(new ArrayList<>(), HttpStatus.OK);
```
Можно было бы не производить создание нового объекта класса `ArrayList`, а использовать готовую пустую неизменяемую коллекцию `List.of()` или `Collections.emptyList()`
Также можно использовать более короткую запись:
`return ResponseEntity.ok(...`

TyutterinYakov commented on gateway/src/main/java/ru/practicum/shareit/user/dto/UserDto.java:
```diff
@@ -0,0 +1,30 @@
+    private String name;
+
+    @NotBlank(groups = Marker.Create.class)
+    @Email
```
Можно было бы указать сразу же две группы:
`@Email(groups = {Marker.OnCreate.class, Marker.OnUpdate.class})`
Аналогично в других местах) Таким образом в контроллере мы будем использовать сразу `@Validated(...)` вместо `@Valid`

TyutterinYakov commented on gateway/src/main/java/ru/practicum/shareit/user/UserController.java:
```diff
@@ -0,0 +1,52 @@
+
+    @PatchMapping("{id}")
+    @Validated({Marker.Update.class})
+    public ResponseEntity<Object> updateUser(@Valid @RequestBody UserDto userDto,
```
После изменений в DTO, здесь мы сможем вместо `@Valid` сразу же использовать `@Validated({Marker.Update.class})`, исключив `@Validated` над классом)

TyutterinYakov commented on gateway/src/main/java/ru/practicum/shareit/user/UserController.java:
```diff
@@ -0,0 +1,52 @@
+
+    @PostMapping
+    @Validated({Marker.Create.class})
+    public ResponseEntity<Object> createUser(@Valid @RequestBody UserDto userDto) {
```
После изменений в DTO, здесь мы сможем вместо `@Valid` сразу же использовать `@Validated({Marker.Create.class})`, исключив `@Validated` над классом)

TyutterinYakov reviewed (commented):
Добрый день! Отметил оставшиеся недочеты) Хорошего дня! Если возникнут вопросы - пиши в пачку, буду рад помочь🌞

sashajaaa committed:
Commit №16.3:
refactor:
-переработан класс gateway/BookingItemRequestDto;
-переработан класс gateway/ItemController;
-переработан класс gateway/UserDto;
-переработан класс gateway/UserController.

sashajaaa committed:
Commit №16.3:
refactor:
-переработан класс gateway/BookingItemRequestDto;
-переработан класс gateway/ItemDto;
-переработан класс gateway/ItemController;
-переработан класс gateway/UserDto;
-переработан класс gateway/UserController.

sashajaaa committed:
Commit №16.3:
refactor:
-переработан класс gateway/BookingItemRequestDto;
-переработан класс gateway/ItemDto;
-переработан класс gateway/ItemController;
-переработан класс gateway/UserDto;
-переработан класс gateway/UserController;
-переработан класс gateway/ItemRequestDto;
-переработан класс gateway/ItemRequestController.

sashajaaa committed:
Commit №16.3:
refactor:
-переработан класс gateway/BookingItemRequestDto;
-переработан класс gateway/ItemDto;
-переработан класс gateway/ItemController;
-переработан класс gateway/UserDto;
-переработан класс gateway/UserController.

sashajaaa committed:
Commit №16.3:
refactor:
-переработан класс gateway/BookingItemRequestDto;
-переработан класс gateway/ItemDto;
-переработан класс gateway/ItemController;
-переработан класс gateway/ItemRequestDto;
-переработан класс gateway/ItemRequestController;
-переработан класс gateway/UserDto;
-переработан класс gateway/UserController.

TyutterinYakov commented on gateway/src/main/java/ru/practicum/shareit/item/ItemController.java:
```diff
@@ -0,0 +1,74 @@
+@Controller
+@RequestMapping(path = "/items")
+@RequiredArgsConstructor
+public class ItemController {
```
Не забываем о `@Validated`, иначе аннотации ограничения по типу `@PositiveOrZero @RequestParam(value = "from", defaultValue = "0") Integer from` и тд - не будут учитываться)

TyutterinYakov reviewed (approved):
Добрый вечер! Поздравляю с успешной сдачей спринта :) Отметил один небольшой недочет, но думаю, что ты справишься с ним вне ревью) Сил и терпения на этом сложном, но интересном пути) Успехов в обучении и прекрасного вечера! Продуктивной работы над дипломным проектом!🥳🌃
